test/src/tests-batchUploader.ts

```diff
@@ -84,16 +84,603 @@ describe('batch uploader', () => {
+                const batchesNotUploaded = await (<any>uploader).uploadBatches(
+                    newLogger,
+                    [batch1, batch2, batch3],
+                    false
+                );
+
+                expect(batchesNotUploaded).to.be.ok;
+
+                expect(
+                    batchesNotUploaded.length,
+                    'Should have 3 uploaded batches'
+                ).to.equal(3);
+
+                expect(
+                    batchesNotUploaded[0].events[0].data.event_name
+                ).to.equal('Test Event 1');
+                expect(
+                    batchesNotUploaded[1].events[0].data.event_name
+                ).to.equal('Test Event 2');
+                expect(
+                    batchesNotUploaded[2].events[0].data.event_name
+                ).to.equal('Test Event 3');
+            });
+        });
+    });
+
+    describe('Upload Workflow', () => {
+        beforeEach(() => {
+            clock = sinon.useFakeTimers({
+                now: new Date().getTime(),
+            });
+        });
+
+        afterEach(() => {
+            window.fetchMock.restore();
+            clock.restore();
+        });
+
+        it('should organize events and batches in the order they are processed and maintain that order when uploading', (done) => {
+            // Batches should be uploaded in the order they were created to prevent
+            // any potential corruption.
+
+            window.fetchMock.post(urls.eventsV3, 200);
+            window.fetchMock.config.overwriteRoutes = true;
+
+            window.mParticle.config.flags = {
+                eventsV3: '100',
+                eventBatchingIntervalMillis: 1000,
+            };
+
+            window.mParticle._resetForTests(MPConfig);
+            window.mParticle.init(apiKey, window.mParticle.config);
+
+            window.mParticle.logEvent('Test Event 0');
+
+            // Force time to pass so that batch processing occurs
+            clock.tick(1000);
+
+            expect(
+                window.fetchMock.called(),
+                'FetchMock should have been called'
+            ).to.equal(true);
+
+            const batch1 = JSON.parse(window.fetchMock._calls[0][1].body);
+
+            // Batch 1 should contain only session start, AST and a single event
+            // in this exact order
+            expect(batch1.events.length).to.equal(3);
+            expect(batch1.events[0].event_type).to.equal('session_start');
+            expect(batch1.events[1].event_type).to.equal(
+                'application_state_transition'
+            );
+            expect(batch1.events[2].data.event_name).to.equal('Test Event 0');
+
+            // Log a second batch of events
+            window.mParticle.logEvent('Test Event 1');
+            window.mParticle.logEvent('Test Event 2');
+            window.mParticle.logEvent('Test Event 3');
+
+            // Force an upload of events
```

Comment:
do a find/replace in this file to change instances of this to what i suggested to be updated elsewhere in the doc: `Manually initiate the upload process - turn event into batches and upload the batch`.